[Security Solution] Add quick job installation to anomalies table

machadoum opened the pull request:
[figma](https://www.figma.com/file/DGMZX8513VFWB8HcAgSPq2/Analytics-Dashboard?node-id=241%3A121014)
[issue](https://github.com/elastic/security-team/issues/5148)


## Summary
Add the 'Run job' button to the anomalies table. The button runs the job and installs it when the module is uninstalled.

**Implementation details:**
* It replaces `useInstalledSecurityJobs` by `useSecurityJobs` to also load uninstalled jobs.
* It updates the error handling code to throw some errors that were previously swallowed.
* Refactor data feed enablement logic to be reusable by making it a hook.
* Refactor `useSecurityJobs(refetchData: boolean)` refetch param from toggle to a callback function.


![Oct-17-2022 10-49-39](https://user-images.githubusercontent.com/1490444/196133232-81982ade-4bf4-49ee-86ef-32d5b0ba5ff7.gif)


### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

machadoum committed:
Add run job button

machadoum committed:
Wip Fix tests

machadoum committed:
Fix issues found during code review

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

machadoum committed:
Add unit tests

machadoum committed:
Remove comments

machadoum commented on x-pack/plugins/security_solution/public/common/components/ml/anomaly/use_anomalies_search.test.ts:
```diff
@@ -254,9 +244,9 @@ describe('useNotableAnomaliesSearch', () => {
           {
             count: 99,
-            jobId: mostRecentJobId,
+            // jobId: mostRecentJobId,
```
Yes. Thank you!

jamster10 commented on x-pack/plugins/security_solution/public/overview/components/entity_analytics/anomalies/columns.tsx:
```diff
@@ -52,13 +50,15 @@ export const useAnomaliesColumns = (loading: boolean): AnomaliesColumns => {
             return -1;
           }
+
           return -2;
```
unrelated to your changes but do you know how are these negative numbers used?

machadoum replied:
Yes, it overrides the column numeric sort order. When the job is uninstalled, we set a negative number to display it at the bottom of the table. So we have the following sort order: 
Jobs with anomalies
Instaled jobs without anomalies
Uninstalled jobs

I can simplify it a bit by using `0` and `-1` instead of `-1` and `-2`

jamster10 commented on x-pack/plugins/security_solution/public/common/components/ml_popover/hooks/use_enable_data_feed.ts:
```diff
@@ -0,0 +1,117 @@
+    }
+    case 'success': {
+      return {
+        ...state,
```
If the State is only ever { isLoading: boolean } why for each case we spread the state? wouldnt it be just the one value, that we end up updating anyway? Also why use a reducer if the state is so simple? Not that it's wrong, just curious

machadoum replied:
Good catch. You are right. We don't have to spread the state. The state used to be more complex, but I deleted some properties in this PR.

I update the code to use `useState`.

jamster10 reviewed (approved):
desk tested and reviewed

machadoum committed:
Fix Anomalies table bug when ES is empty

On an empty ES cluster the anomalies search API response doesn't contain the aggregation

machadoum commented:
@jamster10 I tested that bug you found when the cluster is empty and hopefully it doesn't happen on 8.5. 
So I fixed the bug here and we don't have to backport it.

**Before the Fix:**
<img width="1519" alt="Screenshot 2022-10-19 at 10 45 51" src="https://user-images.githubusercontent.com/1490444/196645057-eeee38ea-7def-4f6b-86d8-5a00714eae4e.png">
**After the fix:**
<img width="1496" alt="Screenshot 2022-10-19 at 10 45 02" src="https://user-images.githubusercontent.com/1490444/196645116-1d8a6038-39da-4e9d-927b-c7980ce93c1e.png">

**How it looks on 8.5**
<img width="1494" alt="Screenshot 2022-10-19 at 10 18 59" src="https://user-images.githubusercontent.com/1490444/196645201-4c9bce29-7385-4667-b994-6681392a3836.png">

machadoum committed:
Code review improvements

angorayc commented on x-pack/plugins/security_solution/public/common/components/ml_popover/hooks/translations.ts:
```diff
@@ -0,0 +1,28 @@
+);
+
+export const STOP_JOB_FAILURE = i18n.translate(
+  'xpack.securitySolution.components.hooks.errors.stopJobFailureTitle',
```
```suggestion
  'xpack.securitySolution.components.mlPopup.hooks.errors.stopJobFailureTitle',
```

machadoum committed:
Update anomalies table load state

machadoum commented:
I updated the loading state to be shared across jobs (same behaviour as ml pop) to fix the issue where the page gets into an inconsistent state when the user starts multiple jobs at once.

![Oct-19-2022 15-14-09](https://user-images.githubusercontent.com/1490444/196701278-a82a4c4a-8c45-4561-9e68-40b85c9aec4a.gif)

machadoum committed:
Fix translation key

banderror commented on x-pack/plugins/security_solution/public/common/components/ml_popover/hooks/use_enable_data_feed.ts:
```diff
@@ -0,0 +1,79 @@
+  const { addError } = useAppToasts();
+  const [isLoading, setIsLoading] = useState(false);
+
+  const enableDatafeed = async (job: SecurityJob, latestTimestampMs: number, enable: boolean) => {
```
I think maybe this should be at least `useCallback`?
I'd suggest though just using React Query which solves the problem of calling external APIs and addresses a ton of edge cases. If you're curious ping us with @xcrzx and we can have a quick chat about what problems can be solved with it. @xcrzx talked about some of that at our On Week demos.

machadoum replied:
Yes, I will use `useCallback` here.

I saw the On Week video and it is very interesting. I will make sure we consider React Query for future security solution refactors. Thanks for sharing it!

banderror commented on x-pack/plugins/security_solution/public/common/components/ml_popover/hooks/use_security_jobs.ts:
```diff
@@ -35,25 +38,24 @@ export interface UseSecurityJobsReturn {
   const http = useHttp();
   const { addError } = useAppToasts();
-
+  const refetch = useRef<inputsModel.Refetch>(noop);
```
For example, refetching is natively supported by React Query and is a matter of simple config for a given hook.

banderror reviewed (approved):
Checked out and tested the `ML job settings` popover on the Rule Management page. Seems to work fine!

I have to admit that I'm not super familiar with the changed code, but I noticed that you've done some refactoring (e.g. `useEnableDataFeed`) and clean-up, and added new tests. Thank you for that @machadoum!

Left some suggestions.

machadoum committed:
Add useCallback to enableDataFeed
